accounts-db/benches/bench_read_only_accounts_cache.rs

```diff
@@ -0,0 +1,208 @@
+        let mut rng = SmallRng::seed_from_u64(100);
+        let mut new_pubkeys: Array2<Pubkey> = Array2::from_shape_vec(
+            (NUM_ACCOUNTS_PER_THREAD / 2, *num_readers_writers),
+            vec![
+                solana_sdk::pubkey::new_rand();
+                *num_readers_writers * (NUM_ACCOUNTS_PER_THREAD / 2)
+            ],
+        )
+        .unwrap();
+        let new_rows: Vec<Vec<Pubkey>> = new_pubkeys
+            .rows()
+            .into_iter()
+            .map(|row| {
+                let mut shuffled_row = row.to_vec();
+                shuffled_row.shuffle(&mut rng);
+                shuffled_row
+            })
+            .collect();
+        for new_row in new_rows {
+            new_pubkeys
+                .push_row(ArrayView::from(new_row.as_slice()))
+                .unwrap();
+        }
+        let new_pubkeys: Vec<Vec<Pubkey>> = new_pubkeys
+            .columns()
+            .into_iter()
+            .map(|column| {
+                // Both `ArrayBase::as_slice` and `ArrayBase::as_mut_slice`
+                // return `None` in this case, so let's just collect the elements.
+                let mut pubkeys_for_thread = column
+                    .into_iter()
+                    .map(|pubkey| pubkey.to_owned())
+                    .collect::<Vec<_>>();
+                pubkeys_for_thread.shuffle(&mut rng);
+                pubkeys_for_thread
+            })
+            .collect();
+
+        // Spawn the reader threads in the background.
+        let stop_reader = Arc::new(AtomicBool::new(false));
+        let reader_handles = (0..*num_readers_writers).map(|i| {
+            let cache = cache.clone();
+            let pubkeys = pubkeys.clone();
+            let stop_reader = stop_reader.clone();
+            Builder::new()
+                .name(format!("reader{i:02}"))
+                .spawn({
+                    move || {
+                        // Continuously read random accounts.
+                        let mut rng = SmallRng::seed_from_u64(i as u64);
+                        while !stop_reader.load(Ordering::Relaxed) {
+                            let pubkey = pubkeys.choose(&mut rng).unwrap();
+                            test::black_box(cache.load(*pubkey, slot));
+                        }
+                    }
+                })
+                .unwrap()
+        });
+
+        // Benchmark reads and writes on a full cache, trigerring eviction on each
+        // write.
+        let slot = 1;
+        group.sample_size(10);
+        group.bench_function(
+            BenchmarkId::new("read_only_accounts_cache_eviction", num_readers_writers),
+            |b| {
+                b.iter(|| {
+                    // Perform the writes.
+                    let writer_handles = (0..*num_readers_writers).map(|i| {
+                        let cache = cache.clone();
+                        let new_pubkeys = new_pubkeys[i].clone();
+
+                        Builder::new()
```

Comment:
I think you should pre-create the writers outside the bench, and make them sleep
on a https://doc.rust-lang.org/stable/std/sync/struct.Barrier.html

Then in the bench you should wait for all the writers to terminate. Otherwise
you're adding thread creation time to your bench, which will get slower with the
more threads you add

Reply:
I actually don't think you can use the barrier, because iter() will want to repeat this many times. You'll need to get a little creative :P

Reply:
ah no nevermind, barriers can be used more than once, nice

Reply:
Yeah, I should also use a `Barrier`/condvar in read threads.

Reply:
Done